plugins/logitech-tap/fu-logitech-tap-sensor-device.c

```diff
@@ -0,0 +1,371 @@
+	return FALSE;
+#endif
+}
+
+static gboolean
+fu_logitech_tap_sensor_device_set_tde(FuDevice *device, const guchar tde_mode, GError **error)
+{
+	FuLogitechTapSensorDevice *self = FU_LOGITECH_TAP_SENSOR_DEVICE(device);
+	guint8 set_data[HID_SET_DATA_LEN] = {kHidMcuTdeReportId, kHidMcuTdeModeSelector, tde_mode, 0, 0};
+
+	/* enable/disable TDE mode */
+	if (!fu_logitech_tap_sensor_device_set_feature(self,
+						      (guchar *)set_data,
+							  HID_SET_DATA_LEN,
+						      error))
+	return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+gboolean
+fu_logitech_tap_sensor_device_reboot_device(FuDevice *device,
+				   GError **error)
+{
+    FuLogitechTapSensorDevice *self = FU_LOGITECH_TAP_SENSOR_DEVICE(device);
+	gint8 pinclr = 05;
+ 	guint8 pinset = 06;
+ 	guint8 PWR = 45;
+ 	guint8 RST = 46;
+ 	guint8 set_data[HID_SET_DATA_LEN] = {kHidReportIdMcuSetCmd, pinclr, PWR, 0, 0};
+    g_autoptr(FuProgress) progress = fu_progress_new(G_STRLOC);
+
+	g_debug("trigger device reboot");
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 100, "attach");
+	fu_progress_set_status(progress, FWUPD_STATUS_DEVICE_RESTART);
+
+	/* need to reopen the device, as at composite_cleaup time, device is already closed */
+	if (!fu_device_open(FU_DEVICE(device), error))
+		return FALSE;
+
+	/* enable TDE mode */
+	if (!fu_logitech_tap_sensor_device_set_tde(device, kHidMcuTdeModeEnable, error))
+		return FALSE;
+
+	/* setup HID report for power cycle */
+	if (!fu_logitech_tap_sensor_device_set_feature(self,
+						      (guchar *)set_data,
+							  HID_SET_DATA_LEN,
+						      error)) {
+		/* disable TDE mode */
+		fu_logitech_tap_sensor_device_set_tde(device, kHidMcuTdeModeDisable, error);
+		return FALSE;
+	}
+	set_data[1] = pinclr;
+    set_data[2] = RST;
+	if (!fu_logitech_tap_sensor_device_set_feature(self,
+						      (guchar *)set_data,
+							  HID_SET_DATA_LEN,
+						      error)) {
+		/* disable TDE mode */
+		fu_logitech_tap_sensor_device_set_tde(device, kHidMcuTdeModeDisable, error);
+		return FALSE;
+	}
+	fu_device_sleep(FU_DEVICE(self), 2000); /* 2 sec */
+    set_data[1] = pinset;
+    set_data[2] = PWR;
+	if (!fu_logitech_tap_sensor_device_set_feature(self,
+						      (guchar *)set_data,
+							  HID_SET_DATA_LEN,
+						      error)) {
+		/* disable TDE mode */
+		fu_logitech_tap_sensor_device_set_tde(device, kHidMcuTdeModeDisable, error);
+		return FALSE;
+	}
+	fu_device_sleep(FU_DEVICE(self), 2000); /* 2 sec */
+    set_data[1] = pinset;
+    set_data[2] = RST;
+	if (!fu_logitech_tap_sensor_device_set_feature(self,
+						      (guchar *)set_data,
+							  HID_SET_DATA_LEN,
+						      error)) {
+
+		/* disable TDE mode */
+		fu_logitech_tap_sensor_device_set_tde(device, kHidMcuTdeModeDisable, error);
+		return FALSE;
+	}
+	/* disable TDE mode */
+	if (!fu_logitech_tap_sensor_device_set_tde(device, kHidMcuTdeModeDisable, error))
+		return FALSE;
```

Comment:
I think it's worth looking into and using a FuDeviceLocker for this, rather than duplicating the TDE cleanup code a ton of times.

Reply:
Thanks for the pointers. Leveraging FuDeviceLocker now